deepchem/models/torch_models/layers.py

```diff
@@ -6150,4 +6150,186 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+    y: tf.Tensor
+        Input Tensor of shape `(m, p)`
+        The shape of this input tensor should be `m` rows by `p` columns.
+        Note that `m` need not equal `n` (the number of rows in `x`).
+
+    Returns
+    -------
+    tf.Tensor
+        Returns a tensor of shape `(n, m)`, that is, `n` rows by `m` columns.
+        Each `i,j`-th entry of this output tensor is the inner product between
+        the l2-normalized `i`-th row of the input tensor `x` and the
+        the l2-normalized `j`-th row of the output tensor `y`.
+    """
+  
+  x_normalized = torch.nn.functional.normalize(x, p=2, dim=1)
+  y_normalized = torch.nn.functional.normalize(x, p=2, dim=1)
+  return torch.matmul(x_normalized, torch.transpose(y_normalized, 1, 0))
+
+class AttnLSTMEmbedding(nn.Module):
+    """Implements AttnLSTM as in matching networks paper.
+    The AttnLSTM embedding adjusts two sets of vectors, the "test" and
+    "support" sets. The "support" consists of a set of evidence vectors.
+    Think of these as the small training set for low-data machine
+    learning.  The "test" consists of the queries we wish to answer with
+    the small amounts of available data. The AttnLSTMEmbdding allows us to
+    modify the embedding of the "test" set depending on the contents of
+    the "support".  The AttnLSTMEmbedding is thus a type of learnable
+    metric that allows a network to modify its internal notion of
+    distance.
+    See references [1]_ [2]_ for more details.
+    References
+    ----------
+    .. [1] Vinyals, Oriol, et al. "Matching networks for one shot learning."
+        Advances in neural information processing systems. 2016.
+    .. [2] Vinyals, Oriol, Samy Bengio, and Manjunath Kudlur. "Order matters:
+        Sequence to sequence for sets." arXiv preprint arXiv:1511.06391 (2015).
+    """
+    def __init__(self, n_test, n_support, n_feat, max_depth, **kwargs):
+        """
+        Parameters
+        ----------
+        n_support: int
+            Size of support set.
+        n_test: int
+            Size of test set.
+        n_feat: int
+            Number of features per atom
+        max_depth: int
+            Number of "processing steps" used by sequence-to-sequence for sets model.
+        """
+        super(AttnLSTMEmbedding, self).__init__(**kwargs)
+        self.max_depth = max_depth
+        self.n_test = n_test
+        self.n_support = n_support
+        self.n_feat = n_feat
+
+        n_feat = self.n_feat
+        self.lstm = LSTMStep(n_feat, 2 * n_feat)
+        self.q_init = torch.zeros([self.n_test, n_feat])
+        self.states_init = self.lstm.get_initial_states([self.n_test, n_feat])
+        self.built = True
+
+    def __repr__(self):
+        return (
+            f'{self.__class__.__name__}(n_test={self.n_test}, n_support={self.n_support}, n_feat={self.n_feat}, max_depth={self.max_depth})'
+        )
+    
+    def forward(self, inputs):
+        """Execute this layer on input tensors.
+        Parameters
```

Comment:
please leave an empty line before this, or it will not render correctly.

Reply:
done